### PR TITLE
Add linters for Docker, Go, Python, and Shell

### PR DESCRIPTION
**Pull request by Jeffrey-Vervoort-KNMI**
~~First, the refactoring of the project needs to be merged (See this branch: https://github.com/EURODEO/e-soh-datastore-poc/tree/generalize_putobservations), and then this PR can be merged.~~ Done

This PR contains:
- Adds linters and formatters for:
   - Docker
   - Go
   - Python
   - Shell
- Adds a github actions for the linters and formatters.
- Adds a pre-commit hook with the linters and formatters.
- Updates the readme on how to install the pre-commit hook.
- Applied the pre-commit hook to all of the files.

### CODE REVIEW

**Comment by Jeffrey-Vervoort-KNMI** [comment]
> First, I think it's great to have these linters in place! I really only have a general comment: how does the pre-commit hook work (in file `.pre-commit-config.yaml`) ? Is source code modified to conform to certain styles? While that's not a problem in itself, it might be challenging if your IDE/editor (vscode in my case) is configured to modify source file according to a _different_ standard upon file save. I.e. the diffs may in certain cases contain a lot of "noise" that will make it harder to distinguish essential changes from automatically applied style changes. But I'm not sure it will work that way in practice, so I'm only guessing here :)

Rebasing was easier on a new branch. Moved Jo's comment [from](https://github.com/EURODEO/e-soh-datastore-poc/pull/43#issuecomment-1742816507) to here and closed the old MR.

**Comment by Jeffrey-Vervoort-KNMI** [comment]
> How does the pre-commit hook work (in file `.pre-commit-config.yaml`) ?

I have updated the README.md in the root of the repository and added the steps to install and work with the pre-commit hook.

> Is source code modified to conform to certain styles? While that's not a problem in itself, it might be challenging if your IDE/editor (vscode in my case) is configured to modify source file according to a _different_ standard upon file save.

For me, it is the reason why I favour linters and formatters. Using them enables anyone to pick the editor they like. Because the tool is then responsible for the formatting and not the editor. Someone can then use VSCode, Intelij or even notepad.

> I.e. the diffs may in certain cases contain a lot of "noise" that will make it harder to distinguish essential changes from automatically applied style changes.

The first time - which is in this MR - there is a lot of noise, thereafter the formatting should be the same for everyone. This should minimise the diffs.

**Comment by lukas-phaf** [comment]
I think what Jo is worried about is formatting "ping-pong". The editor saves with "format A", the pre-commit hooks run, reformat the code to "format B", the editor puts it back to "format A", etc.

But maybe we should just try it? As this MR adds only a single space to the Go code, and we have no issues with this in Python (although not sure if anybody uses vscode in our group) with these linters.
We can also back out if we see problems.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -7,7 +7,87 @@ on:
     tags:
       - '[0-9]+.[0-9]+.[0-9]+'
 
+permissions:
+  contents: read
+
+env:
+  LINE_LENGTH: 99  # TODO: would we like to increase this to 120?
```
Comment (lukas-phaf) [inline .github/workflows/ci.yml:14]: Yes, please!

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -7,7 +7,87 @@ on:
...
+
 jobs:
+  docker-lint-hadolint:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: hadolint/hadolint-action@v3.1.0
+        with:
+          ignore: DL3008
+          recursive: true
```
Comment (lukas-phaf) [inline .github/workflows/ci.yml:24]: Seperate jobs which each do a checkout, etc, versus one job where all the different linting is done in separate steps.

Or see if we can just run the pre-commit hooks here instead using:
`pre-commit run --config './.pre-commit-config.yaml' --all-files`

File: tstester/config.json [hunk 3]
```diff
@@ -1,39 +1,31 @@
...
-    "_comment": "probability weights of time series resolutions (around 20% will have time res 60 secs, around 30% will have time res 600 secs, and so on)",
-
-    "extra_secs": [60, 600, 3600],
-    "_comment": "extra secs values to try with the AddNewObs test",
-
+    "extra_secs": [
+        60,
+        600,
+        3600
+    ],
```
Comment (lukas-phaf) [inline tstester/config.json:24]: Keep all the comments by renaming them to _comment1, _comment2, etc.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,15 @@
 # e-soh-datastore-poc
 E-SOH datastore PoCs
+
+
+## Pre-commit
+### Setup
+1. Go to the root of the repository.
+2. Install the python pre-commit package with `pip install pre-commit`.
```
[the remark this answers is not in the document]
Comment (Jeffrey-Vervoort-KNMI) [reply README.md:8]: I have the Python pre-commit installed in my global environment. I like to use it in every project. So I would say no.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Jeffrey-Vervoort-KNMI) [reply .github/workflows/ci.yml]: Applied

File: .github/workflows/ci.yml (hunk 2 above)
Comment (Jeffrey-Vervoort-KNMI) [reply .github/workflows/ci.yml]: We can try this out, applied it.

File: tstester/config.json (hunk 3 above)
Comment (Jeffrey-Vervoort-KNMI) [reply tstester/config.json]: Well seen, I did not notice that the pre-commit hook removed some of the comments.

Comment (lukas-phaf) [reply tstester/config.json]: It doesn't accept duplicates... :-)

**Review by lukas-phaf — APPROVED** [review]

**Pull request merged**